File: iosApp/iosApp/Presentation/Cast/CastViewModel.swift
//
//  CastViewModel.swift
//  ConCafe
//
//  Created by 홍희표 on 2026/03/06.
//

import Foundation
import Combine
import Shared
import KMPNativeCoroutinesAsync

@MainActor
final class CastViewModel: ObservableObject {
    private let castId: String

    private let getCastDetailUseCase: GetCastDetailUseCase

    private let toggleFollowCastUseCase: ToggleFollowCastUseCase

    private let castEventPublisher: CastEventPublisher

    private let reviewEventPublisher: ReviewEventPublisher

    @Published private(set) var uiState = CastUiState.empty

    let event = PassthroughSubject<CastEvent, Never>()

    private var tasks: [TaskKey: Task<Void, Never>] = [:]

    private func observeCastEvent() {
        tasks[.castEvent]?.cancel()
        tasks[.castEvent] = Task {
            do {
                for try await event in asyncSequence(for: castEventPublisher.events) {
                    switch event {
                    case let event as Shared.CastEvent.Created:
                        if event.cast.id == self.castId {
                            self.loadCastDetail()
                        }
                    case let event as Shared.CastEvent.Updated:
                        if event.cast.id == self.castId, let detail = self.uiState.detail {
                            self.uiState.detail = CastDetail(
                                cast: event.cast,
                                cafe: detail.cafe,
                                images: detail.images,
                                schedule: detail.schedule,
                                visitCertificationCount: detail.visitCertificationCount
                            )
                        }
                    case let event as Shared.CastEvent.Deleted:
                        if event.castId == self.castId {
                            self.event.send(.navigateBack)
                        }
                    default:
                        break
                    }
                }
            } catch {
                print("Error: \(error)")
            }
        }
    }

    private func observeReviewEvent() {
        tasks[.reviewEvent]?.cancel()
        tasks[.reviewEvent] = Task {
            do {
                for try await event in asyncSequence(for: reviewEventPublisher.events) {
                    guard let currentCafeId = self.uiState.detail?.cafe.id else { return }
                    if let created = event as? ReviewEvent.Created, created.cafeId == currentCafeId {
                        self.loadCastDetail()
                    } else if let deleted = event as? ReviewEvent.Deleted, deleted.cafeId == currentCafeId {
                        self.loadCastDetail()
                    }
                }
            } catch {
                print("Error: \(error)")
            }
        }
    }

    private func loadCastDetail() {
        uiState.isLoading = true
        uiState.errorMessage = nil

        Task {
            do {
                let result = try await getCastDetailUseCase.invoke(castId: castId)

                if let success = result as? AppResultSuccess<AnyObject>,
                   let feed = success.data as? Shared.CastDetailFeed {
                    uiState = CastUiState(
                        isLoading: false,
                        errorMessage: nil,
                        detail: feed.detail,
                        recentReviews: feed.recentReviews,
                        isFollowing: feed.isFollowing,
                        isLoggedIn: feed.isLoggedIn,
                        todayAttendanceStatus: feed.todayAttendanceStatus,
                        isSelfCast: feed.isSelfCast
                    )
                } else {
                    uiState.isLoading = false
                    uiState.errorMessage = nil
                }
            } catch {
                if Task.isCancelled { return }
                uiState.isLoading = false
                uiState.errorMessage = nil
            }
        }
    }

    private func toggleFollow() {
        Task {
            do {
                let result = try await toggleFollowCastUseCase.invoke(castId: castId)

                if let success = result as? AppResultSuccess<AnyObject>,
                   let isFollowing = success.data as? NSNumber {
                    uiState.isFollowing = isFollowing.boolValue
                    uiState.isLoggedIn = true
                } else if let failure = result as? AppResultFailure {
                    if failure.error is AppErrorUnauthorized {
                        event.send(.navigateToSignIn)
                    }
                }
            } catch {
                if Task.isCancelled { return }
            }
        }
    }

    func onAction(_ action: CastAction) {
        switch action {
        case .backTapped:
            event.send(.navigateBack)
        case .followTapped:
            toggleFollow()
        case .refresh:
            loadCastDetail()
        case .cafeTapped:
            guard let cafeId = uiState.detail?.cafe.id else { return }
            event.send(.navigateToCafe(id: cafeId))
        }
    }

    init(
        castId: String,
        getCastDetailUseCase: GetCastDetailUseCase = KoinInitializerKt.resolveGetCastDetailUseCase(),
        toggleFollowCastUseCase: ToggleFollowCastUseCase = KoinInitializerKt.resolveToggleFollowCastUseCase(),
        castEventPublisher: CastEventPublisher = KoinInitializerKt.resolveCastEventPublisher(),
        reviewEventPublisher: ReviewEventPublisher = KoinInitializerKt.resolveReviewEventPublisher()
    ) {
        self.castId = castId
        self.getCastDetailUseCase = getCastDetailUseCase
        self.toggleFollowCastUseCase = toggleFollowCastUseCase
        self.castEventPublisher = castEventPublisher
        self.reviewEventPublisher = reviewEventPublisher

        observeCastEvent()
        observeReviewEvent()
        loadCastDetail()
    }

    deinit {
        tasks.values.forEach { $0.cancel() }
        tasks.removeAll()
    }

    private enum TaskKey {
        case castEvent
        case reviewEvent
    }
}

File: composeApp/src/commonMain/kotlin/com/hhp227/concafe/presentation/cast/CastViewModel.kt
package com.hhp227.concafe.presentation.cast

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import com.hhp227.concafe.domain.common.AppError
import com.hhp227.concafe.domain.common.AppResult
import com.hhp227.concafe.domain.event.CastEvent as CastDomainEvent
import com.hhp227.concafe.domain.event.ReviewEvent
import com.hhp227.concafe.domain.event.publisher.CastEventPublisher
import com.hhp227.concafe.domain.event.publisher.ReviewEventPublisher
import com.hhp227.concafe.domain.usecase.GetCastDetailUseCase
import com.hhp227.concafe.domain.usecase.ToggleFollowCastUseCase

class CastViewModel(
    private val castId: String,
    private val getCastDetailUseCase: GetCastDetailUseCase,
    private val toggleFollowCastUseCase: ToggleFollowCastUseCase,
    private val castEventPublisher: CastEventPublisher,
    private val reviewEventPublisher: ReviewEventPublisher
) : ViewModel() {
    private val _uiState = MutableStateFlow(CastUiState.empty())
    val uiState: StateFlow<CastUiState> = _uiState.asStateFlow()

    private val _event = MutableSharedFlow<CastEvent>(replay = 0)
    val event = _event.asSharedFlow()

    private fun observeCastEvent() {
        viewModelScope.launch {
            castEventPublisher.events.collectLatest { event ->
                when (event) {
                    is CastDomainEvent.Created -> if (event.cast.id == castId) {
                        loadCastDetail()
                    }
                    is CastDomainEvent.Updated -> if (event.cast.id == castId) {
                        _uiState.update { state ->
                            state.copy(detail = state.detail?.copy(cast = event.cast))
                        }
                    }
                    is CastDomainEvent.Deleted -> if (event.castId == castId) {
                        _event.emit(CastEvent.NavigateBack)
                    }
                }
            }
        }
    }

    private fun observeReviewEvent() {
        viewModelScope.launch {
            reviewEventPublisher.events.collectLatest { event ->
                val currentCafeId = _uiState.value.detail?.cafe?.id ?: return@collectLatest
                when (event) {
                    is ReviewEvent.Created -> if (event.cafeId == currentCafeId) {
                        loadCastDetail()
                    }
                    is ReviewEvent.Deleted -> if (event.cafeId == currentCafeId) {
                        loadCastDetail()
                    }
                }
            }
        }
    }

    private fun loadCastDetail() {
        _uiState.update {
            it.copy(
                isLoading = true,
                errorMessage = null
            )
        }

        viewModelScope.launch {
            when (val result = getCastDetailUseCase.invoke(castId)) {
                is AppResult.Success -> {
                    _uiState.value = CastUiState(
                        isLoading = false,
                        errorMessage = null,
                        detail = result.data.detail,
                        recentReviews = result.data.recentReviews,
                        isFollowing = result.data.isFollowing,
                        isLoggedIn = result.data.isLoggedIn,
                        todayAttendanceStatus = result.data.todayAttendanceStatus,
                        isSelfCast = result.data.isSelfCast
                    )
                }
                is AppResult.Failure -> {
                    _uiState.update {
                        it.copy(
                            isLoading = false,
                            errorMessage = "캐스트 상세 데이터를 불러오지 못했습니다."
                        )
                    }
                }
            }
        }
    }

    private fun toggleFollow() {
        viewModelScope.launch {
            when (val result = toggleFollowCastUseCase.invoke(castId)) {
                is AppResult.Success -> {
                    _uiState.update {
                        it.copy(
                            isFollowing = result.data,
                            isLoggedIn = true
                        )
                    }
                }
                is AppResult.Failure -> {
                    if (result.error is AppError.Unauthorized) {
                        _event.emit(CastEvent.NavigateToSignIn)
                    }
                }
            }
        }
    }

    fun onAction(action: CastAction) {
        viewModelScope.launch {
            when (action) {
                CastAction.ClickBack -> _event.emit(CastEvent.NavigateBack)
                CastAction.ClickFollow -> toggleFollow()
                CastAction.Refresh -> loadCastDetail()
                CastAction.ClickCafe -> {
                    val cafeId = _uiState.value.detail?.cafe?.id ?: return@launch
                    _event.emit(CastEvent.NavigateToCafe(cafeId))
                }
            }
        }
    }

    init {
        observeCastEvent()
        observeReviewEvent()
        loadCastDetail()
    }
}
